### PR TITLE
enhance load xclbin flow for CU index and more

### PR DESCRIPTION
**Pull request by mamin506**
In the new KDS, the CU index is determined in driver.

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 1]
```diff
@@ -1347,12 +1347,72 @@ int shim::xclLoadXclBin(const xclBin *buffer)
...
+     * |   anums               |
+     * |   argument[0]         |
+     * |   argument[1]         |
+     * |   argument[...]       |
+     * |-----------------------|
+     * | Kernel[...]           |
+     * |   ...                 |
+     * +-----------------------+
+     */
+    axlf_obj.kernels = (char *)malloc(axlf_obj.ksize);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1390]: Please no malloc/free/new/delete.  Same comment as earlier

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 2]
```diff
@@ -581,13 +581,50 @@ get_kernel_arguments(const char* xml_data, size_t xml_size, const std::string& k
...
 }
 
+std::vector<kernel_object>
+get_kernels(const axlf* top)
+{
+  auto xml = get_xml_section(top);
+  std::vector<kernel_object> kernels;
+
+  auto knames = get_kernel_names(xml.first, xml.second);
+  for (auto kname : knames) {
```
Comment (stsoe) [inline src/runtime_src/core/common/xclbin_parser.cpp:618]: auto&

File: src/runtime_src/core/edge/user/shim.cpp [hunk 3]
```diff
@@ -508,7 +509,38 @@ xclLoadAxlf(const axlf *buffer)
   drm_zocl_axlf axlf_obj = {
     .za_xclbin_ptr = const_cast<axlf *>(buffer),
     .za_flags = flags,
+    .za_ksize = 0,
+    .za_kernels = NULL,
   };
+
+  auto kernels = xrt_core::xclbin::get_kernels(buffer);
+  /* Calculate size of kernels */
+  for (auto kernel : kernels) {
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:518]: auto&

File: src/runtime_src/core/edge/user/shim.cpp [hunk 4]
```diff
@@ -508,7 +509,38 @@ xclLoadAxlf(const axlf *buffer)
...
+      axlf_obj.za_ksize += sizeof(kernel_info) + sizeof(argument_info) * kernel.args.size();
+  }
+
+  /* Check PCIe's shim.cpp for details of kernels binary */
+  axlf_obj.za_kernels = (char *)malloc(axlf_obj.za_ksize);
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:523]: Please no malloc/free/new/delete, this is not exception safe
Use std::vector<char*> data(size) to manage the buffer and extract the buffer where necessary using vec.data().

But I am not seeing a free of this buffer.  Wouldn't it be better if the driver allocated the memory and did the copying?   A dangling malloc is not acceptable.  You already have all the data allocated on the user side, so the copying done here could as well be done in the driver.  You can define a struct like kernel_info that just contains the user pointer addresses already allocated. e.g. kernel.name.c_str() which is valid for the duration of the xclLoad call so no additional strcpy is necessary.

File: src/runtime_src/core/edge/user/shim.cpp [hunk 5]
```diff
@@ -508,7 +509,38 @@ xclLoadAxlf(const axlf *buffer)
...
+  for (auto kernel : kernels) {
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:524]: auto&

File: src/runtime_src/core/edge/user/shim.cpp [hunk 6]
```diff
@@ -508,7 +509,38 @@ xclLoadAxlf(const axlf *buffer)
...
+      auto krnl = reinterpret_cast<kernel_info *>(axlf_obj.za_kernels + off);
+      strcpy(krnl->name, kernel.name.c_str());
+      krnl->anums = kernel.args.size();
+
+      int ai = 0;
+      for (auto arg : kernel.args) {
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:530]: auto&

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 7]
```diff
@@ -1347,12 +1347,72 @@ int shim::xclLoadXclBin(const xclBin *buffer)
 int shim::xclLoadAxlf(const axlf *buffer)
 {
     xrt_logmsg(XRT_INFO, "%s, buffer: %s", __func__, buffer);
+    drm_xocl_axlf axlf_obj = {const_cast<axlf *>(buffer), 0};
+    int off = 0;
+
+    auto kernels = xrt_core::xclbin::get_kernels(buffer);
+    /* Calculate size of kernels */
+    for (auto kernel : kernels) {
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1355]: auto&

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 8]
```diff
@@ -1347,12 +1347,72 @@ int shim::xclLoadXclBin(const xclBin *buffer)
...
+     * |   argument[0]         |
+     * |   argument[1]         |
+     * |   argument[...]       |
+     * |-----------------------|
+     * | Kernel[...]           |
+     * |   ...                 |
+     * +-----------------------+
+     */
+    axlf_obj.kernels = (char *)malloc(axlf_obj.ksize);
+    for (auto kernel : kernels) {
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1391]: auto&

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 9]
```diff
@@ -1347,12 +1347,72 @@ int shim::xclLoadXclBin(const xclBin *buffer)
...
+        auto krnl = reinterpret_cast<kernel_info *>(axlf_obj.kernels + off);
+        strcpy(krnl->name, kernel.name.c_str());
+        krnl->anums = kernel.args.size();
+
+        int ai = 0;
+        for (auto arg : kernel.args) {
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1397]: auto&

**Comment by maxzhen** [comment]
> 
> 
> Cool. Lots of changes :-). I added some requests for changes.
> 
> A general question: User land does not use the ordering determined by the driver, in other words, xclbin_parser::get_cus() is still computing the order? We want to change this, right? Not needed now of course.

Hi @stsoe , yes, we want to change this so that xclbin_parser::get_cus() can read all CU information from driver including the index. @mamin506 , if the order you have in driver is different than in the get_cus(), we will have to change it now. I think that the cu order in user space and driver should be in sync.

**Comment by mamin506** [comment]
> Hi @stsoe , yes, we want to change this so that xclbin_parser::get_cus() can read all CU information from driver including the index. @mamin506 , if the order you have in driver is different than in the get_cus(), we will have to change it now. I think that the cu order in user space and driver should be in sync.

The order is the same. But current get_cus() still required by mb_scheduler. We would definitely change this function and get CU information from driver.

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 9 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/linux/shim.cpp]: fixed

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 8 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/linux/shim.cpp]: fixed

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 7 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/linux/shim.cpp]: fixed

File: src/runtime_src/core/edge/user/shim.cpp (hunk 6 above)
Comment (mamin506) [reply src/runtime_src/core/edge/user/shim.cpp]: fixed

File: src/runtime_src/core/edge/user/shim.cpp (hunk 5 above)
Comment (mamin506) [reply src/runtime_src/core/edge/user/shim.cpp]: fixed

File: src/runtime_src/core/edge/user/shim.cpp (hunk 4 above)
Comment (mamin506) [reply src/runtime_src/core/edge/user/shim.cpp]: yes, you are right. Driver would allocate a buffer and copy this binary.
Use c++ container is much more better. :)

File: src/runtime_src/core/edge/user/shim.cpp (hunk 3 above)
Comment (mamin506) [reply src/runtime_src/core/edge/user/shim.cpp]: fixed

File: src/runtime_src/core/common/xclbin_parser.cpp (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/core/common/xclbin_parser.cpp]: fixed

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 1 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/linux/shim.cpp]: fixed

**Pull request merged**